packages/js/src/utils/generators.ts

```diff
@@ -0,0 +1,58 @@
+/** @link [Proof Key for Code Exchange by OAuth Public Clients](https://datatracker.ietf.org/doc/html/rfc7636) */
+
+import { UrlSafeBase64 } from '@silverhand/essentials';
+import { fromUint8Array } from 'js-base64';
+import { customAlphabet } from 'nanoid';
+
+/**
+ * Characters Range of Code Verifier
+ *
+ * The code verifier is a string only containing the unreserved characters [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~" ( url-safe ) .
+ * @link [Client Creates a Code Verifier](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1)
+ */
+export const CODE_VERIFIER_ALPHABET =
+  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+
+/**
+ * Length Range of Code Verifier
+ *
+ * The length of code verifier ranges from 43 to 128.
+ * @link [Client Creates a Code Verifier](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1)
+ */
+export const CODE_VERIFIER_MAX_LENGTH = 128;
+
+/**
+ * @param length The length of the generated string.
+ */
+function generateRandomString(length: number) {
+  return customAlphabet(CODE_VERIFIER_ALPHABET, length)();
+}
+
+/**
+ * Generates random bytes for state and encodes them in url safe base64
+ */
+export const generateState = () => generateRandomString(CODE_VERIFIER_MAX_LENGTH);
+
+/**
+ * Generates random bytes for nonce and encodes them in url safe base64
+ */
+export const generateNonce = () => generateRandomString(CODE_VERIFIER_MAX_LENGTH);
```

Comment:
Do we cover this method?  Looks like missing from the conversion doc

Reply:
I'm good with pre-defining unused utility function for now.
(add it according to the implementation of `oidc-client`)

@xiaoyijun any ideas?

- A. remove for now
- B. reserve it

Reply:
A